### ISSUE TITLE
[Question] Proposal for WebAPI calls without JS glue

### ISSUE THREAD

**Issue by munrocket** [WebAssembly/component-model#371]
I am lost with current proposals. As I know Interface types proposal was created for direct WebAPI calls from WebAssembly. (Answer from Emscripten team https://github.com/emscripten-core/emscripten/issues/15710#issuecomment-987263141 )

But it was renamed to component model and slowed down? How WebAssembly WebGL applications supposed to be faster than pure JS?

**Comment by lukewagner** [comment]
The goals of this proposal have shifted along with the name of the proposal as our understanding of the problem space has improved and the usage of wasm has expanded beyond the browser.  E.g., there were some experiments a few years back by the Emscripten team showing that, at least for the WebGL workloads that they were measuring, the actual overhead of JS glue code wasn't significant enough to motivate continued work on what was then called "Web IDL Bindings".  While I think the component model could indeed still be used to optimize away JS glue code when calling from wasm into the browser, that's no longer the main driving goal for the C-M (see [Goals.md](https://github.com/WebAssembly/component-model/blob/main/design/high-level/Goals.md) and [UseCases.md](https://github.com/WebAssembly/component-model/blob/main/design/high-level/UseCases.md) for what are).  In the meantime, better browser integration has been happening incrementally both in core wasm (with the addition of `externref`) and the JS API (see [js-promise-integration](https://github.com/WebAssembly/js-promise-integration) and [js-string-builtins](https://github.com/WebAssembly/js-string-builtins)), all of which can be used to optimize the generated glue code by the Component Model's browser polyfill, [`jco`](https://github.com/bytecodealliance/jco).  Based on the last in-person CG meeting, I expect there will be further incremental optimizations motivated by optimizing away significant sources of overhead, especially copying of data into and out of streams and WebGPU buffers.

**Comment by anuoua** [comment]
> The goals of this proposal have shifted along with the name of the proposal as our understanding of the problem space has improved and the usage of wasm has expanded beyond the browser. E.g., there were some experiments a few years back by the Emscripten team showing that, at least for the WebGL workloads that they were measuring, the actual overhead of JS glue code wasn't significant enough to motivate continued work on what was then called "Web IDL Bindings". While I think the component model could indeed still be used to optimize away JS glue code when calling from wasm into the browser, that's no longer the main driving goal for the C-M (see [Goals.md](https://github.com/WebAssembly/component-model/blob/main/design/high-level/Goals.md) and [UseCases.md](https://github.com/WebAssembly/component-model/blob/main/design/high-level/UseCases.md) for what are). In the meantime, better browser integration has been happening incrementally both in core wasm (with the addition of `externref`) and the JS API (see [js-promise-integration](https://github.com/WebAssembly/js-promise-integration) and [js-string-builtins](https://github.com/WebAssembly/js-string-builtins)), all of which can be used to optimize the generated glue code by the Component Model's browser polyfill, [`jco`](https://github.com/bytecodealliance/jco). Based on the last in-person CG meeting, I expect there will be further incremental optimizations motivated by optimizing away significant sources of overhead, especially copying of data into and out of streams and WebGPU buffers.

In fact, many people want to develop web with the original wasm language. I originally planned to write a wasm front-end framework, but I felt it was extremely bad when I thought about not being able to directly call various browser APIs and having to do a lot of gluing work. The glue increases the complexity, makes debugging more difficult, unintuitive, and loses performance. Personally, I can't accept that wasm can't directly operate the API.

**Comment by qouteall** [comment]
In this article [Why is WebAssembly a second-class language on the web?](https://hacks.mozilla.org/2026/02/making-webassembly-a-first-class-language-on-the-web/) it says that the component model aim to make Wasm call API without JS glue. It also mentions the 45% peformance loss of DOM API of JS glue. So it seems that the goal has changed?

**Comment by eggyal** [comment]
(cc @eqrion as author of the linked article)

**Comment by eqrion** [comment]
@qouteall Yes, I'm personally interested in using the Component Model to support direct Wasm->Web API calls. From Luke's comment above, it's not that this wasn't a goal, it just wasn't the most important thing. There are other benefits from Components, and different ecosystems (serverless vs browser) value different things. That article is written just from the browser's perspective.

On potential performance gains: we're going to be looking into this more again. Any performance gain is very workload dependent. Carefully constructed WebGL API calls can be just as fast in wasm as in JS. DOM API's tend to require many more data conversions when wasm can't call the API's directly, and therefore can see more of a boost.

**Comment by eggyal** [comment]
@eqrion Is this something that you're just looking at internally within Mozilla, or would there be some value in coordinating a community effort (offline from this GitHub issue/repo)?  I'd be interested in getting involved.

**Comment by eqrion** [comment]
We're looking at doing some prototyping of a subset of the CM to test out API performance. We hope to have it behind a preference in Firefox so that others could experiment with it as well. So at that point it could become more of a community effort.

**Comment by kripken** [comment]
About potential performance gains, it would also be good to measure the [batching](https://github.com/io-eric/webcc/blob/main/docs/architecture.md#architecture) approach of [webcc](https://github.com/io-eric/webcc). Their numbers show [more than a 2x speedup](https://github.com/io-eric/webcc?tab=readme-ov-file#benchmarks), which makes sense as they basically remove the boundary cost entirely, at least in cases where batches are not constantly flushed.

**Comment by kripken** [comment]
Following the last comment, https://github.com/WebAssembly/binaryen/pull/8530 has an experiment with batching Web API calls on the toolchain side. This is not 100% realistic (it is not fully optimized on the one hand, and on the other won't work out of the box on some APIs) but it shows large fps speedups, **1.5x** and **2x** on cases where wasm->js call overhead is significant. The overhead of JS glue basically goes away if you can batch enough calls.

(I don't have immediate plans to polish and land this, because for the use cases I'm most aware of - apps using WebGL/WebGPU to render - wasm->js call overhead isn't an issue. But if this would be useful for other use cases or toolchains I'd be happy to do the work.)

**Comment by ApoloApps** [comment]
I've been benchmarking [Compose Multiplatform for Web](https://github.com/JetBrains/compose-multiplatform) (uses a Canvas with Wasm and manages the ui by itself) and the most time spent in traces is the boundary between wasm2js. Not only does it use WebGL apis but any browser api that requires the browser to provide any information. The framework would greatly benefit from this!

**Comment by kripken** [comment]
@ApoloApps Is there a link to a build of that benchmark? I'd be curious to test it.

**Comment by MaxGraey** [comment]
>  E.g., there were some experiments a few years back by the Emscripten team showing that, at least for the WebGL workloads that they were measuring, the actual overhead of JS glue code wasn't significant enough to motivate continued work on what was then called "Web IDL Bindings"

@lukewagner 

According to your presentation in 2019 and benchmark screenshot, the difference in performance was quite significant (up to 5x). Why did Mozilla team decide to discontinue further work in this area and change direction compleately?
https://vimeo.com/350135351

<img width="951" height="530" alt="Image" src="https://github.com/user-attachments/assets/22b0b4d5-822f-419b-8924-4670546d94ff" />

**Comment by lukewagner** [comment]
@MaxGraey The Emscripten team I was referring to in that quoted comment wasn't at Mozilla, and I don't know the actual benchmarks that were used to evaluate.  I expect it was a WebGL workload and along the lines of what Alon is saying in that, if you change the structure of your code to batch calls across the wasm-to-JS boundary, you can significantly reduce the overhead of crossing the boundary.  From a "today" perspective, I think it'd be great for wasm adoption if one didn't have to optimize around this constraint.

**Comment by anuoua** [comment]
According to the first principle thinking, WASM should directly call the Web API.

**Comment by sagittarius255** [comment]
According to Bytecode Alliance, the component model cannot reach 1.0 status until we get native DOM access from WebAssembly in at least two browsers without using JavaScript:

https://bytecodealliance.org/articles/the-road-to-component-model-1-0

So at this point it's really a question of if and when browsers like Chrome and Firefox can implement the component model and native DOM access from WebAssembly.

**Comment by sagittarius255** [comment]
Direct DOM access from WebAssembly also does not mean that we will be able to get rid of JavaScript fully from the frontend because JavaScript polyfills are still needed for backwards compatibility:

https://www.webassembly-wasm.com/webassembly-core-concepts-browser-runtime/polyfill-alternatives-fallbacks/

such as if you're stuck on an operating system like Windows 7 with last supported Chromium version 109.
